microprofile/WFLY-12313_MicroProfile_OpenAPI_1.1.adoc

```diff
@@ -0,0 +1,126 @@
+= Integrate MicroProfile OpenAPI 1.1
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This enhancement proposes to add support for https://download.eclipse.org/microprofile/microprofile-open-api-1.1.2/microprofile-openapi-spec.html[Eclipse MicroProfile OpenAPI 1.1] to WildFly by integrating the implementation from https://github.com/smallrye/smallrye-open-api[SmallRye].
+This spec provides a unified Java API for producing documentation that complies with the https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md[OpenAPI v3 specification].
+
+An MicroProfile OpenAPI subsystem exposes an HTTP GET `/openapi` endpoint that returns an OpenAPI v3 compliant document (in YAML or JSON format) containing documentation of the RESTful services of an application.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12313
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+The SmallRye implementation of the MicroProfile OpenAPI specification is already integrated by the following projects:
+
+* https://thorntail.io/[Thorntail] via the https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-openapi[microprofile-openapi fraction].
+* https://quarkus.io/[Quarkus] via the https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-openapi[smallrye-openapi extension].
+
+These integrations will heavily inform, though not dictate, the corresponding integration into WildFly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* Create an MicroProfile OpenAPI galleon layer containing the MP Open API feature encapsulated by a new MicroProfile OpenAPI subsystem.
+** Should depend on MicroProfile Config and JAX-RS (and their dependencies)
+** N.B. New modules to be included in WF:
+*** MicroProfile OpenAPI subsystem module
+*** io.smallrye:smallrye-open-api
+*** org.eclipse.microprofile.openapi:microprofile-openapi-api
+*** com.fasterxml.jackson.dataformat:jackson-dataformat-yaml
+** Which feature packs will use this layer is still TBD.
```

Comment:
We should have answers for these questions in the coming weeks.

Reply:
FYI - we will add microprofile-openapi to the default standalone-openshift.xml config.  I'll update the document.

Reply:
Only that we will provide a galleon layer for microprofile-openapi and that the subsystem will be included in the default standalone-openshift.xml.
As such, our general documentation will need to include adding the extension to the target server profile as a first step.

Reply:
> Only that we will provide a galleon layer for microprofile-openapi 

Ok, this one is covered in the proposal.

> ...the subsystem will be included in the default standalone-openshift.xml.

And this is probably where to add to the document. If so, will the feature be available by default _only_ in `standalone` operating mode and not `domain`, right? 
That said, then is there any limitation for the users to later enable the subsystem when operating Wildfly in both `standalone` and `domain` modes?

> As such, our general documentation will need to include adding the extension to the target server profile as a first step.

Ok. Will it be available for each profile then? Anyway, I suppose you're referring to community docs here and I agree in such case.
However the proposal should provide detailed answers to previous questions so that is clear from this very document how actually the feature will be available for end user cases.

Reply:
> If so, will the feature be available by default only in standalone operating mode and not domain, right?

Yes. Domain mode is not relevant to OpenShift.

> That said, then is there any limitation for the users to later enable the subsystem when operating Wildfly in both standalone and domain modes?

No.  I am not aware of any subsystem with such a limitation.

> Will it be available for each profile then?

Any extension is available can be added to any profile.  There are no inherent limitations.